shared/submissionRunner/types.cts

```diff
@@ -0,0 +1,36 @@
+import { z } from 'zod';
+
+const RunResultKind = z.enum([
+	'CompileFailed',
+	'TimeLimitExceeded',
+	'Completed',
+	'SandboxError',
+	'RunError'
+]);
+
+export type RunResultKind = z.infer<typeof RunResultKind>;
+
+export const RunResultZod = z
+	.object({
+		kind: RunResultKind,
+		output: z.string().optional(),
+		exitCode: z.number().optional(),
+		runtimeMilliseconds: z.number().optional(),
+		resultKindReason: z.string().optional()
+	})
+	.strict();
+
+export type RunResult = z.infer<typeof RunResultZod>;
+
+export interface IRunnerParams {
+	srcDir: string;
+	input: string;
+	outputCallback?: (data: string) => void;
+}
+
+export type IRunnerReturn =
+	{ success: true; killFunc: () => void; runResult: Promise<RunResult> } |
+	{ success: false; runResult: RunResult };
+
+export type IRunner<T extends IRunnerParams = IRunnerParams> = 
+	(params: T) => Promise<IRunnerReturn>;
```

Comment:
It's a bit lost in this PR, but I had to change these types. 

The `IRunner<...>` used to be a method that returned `IRunnerReturn`, which was a type expressed as `Promise< {success case} | {failure case} >`.

That structure threw errors in this restructured shared project, so I changed it to have the `IRunner<...>` method to return a `Promise<IReturnRunner>` (which makes sense, it's a method you should await and you can see that more clearly on the signature now), with `IRunnerReturn` just directly `{success case} | {failure case}` (not wrapped in a `Promise`, which also makes sense).

The old structure must've been fine in one `tsconfig` setup, but this one works here and also is easier to read IMO.

(The change of `IRunner<...>` from an interface to type was personal preference, I find the interface-with-one-unnamed-method pattern weird, and some tools recommend switching to this as well. Let me know if you really prefer it the previous way.)

Reply:
Seems good to be, I agree the previous types were pretty ugly